src/Shared/UnitTests/FileMatcher_Tests.cs

```diff
@@ -1374,6 +1374,90 @@ public void FileEnumerationCacheTakesExcludesIntoAccount()
+
+                try
+                {
+                    (string[] fileList, FileMatcher.SearchAction action, string excludeFileSpec) = FileMatcher.Default.GetFiles(
+                        string.Empty,
+                        NativeMethodsShared.IsWindows ? @"c:\**\*.cs" : "/**/*.cs");
+
+                    action.ShouldBe(FileMatcher.SearchAction.FailOnDriveEnumeratingWildcard);
+                    fileList.ShouldBeEmpty();
+                    excludeFileSpec.ShouldBe(string.Empty);
+
+                    (fileList, action, excludeFileSpec) = FileMatcher.Default.GetFiles(
+                        string.Empty,
+                        NativeMethodsShared.IsWindows ? @"c:\**" : "/**");
+
+                    action.ShouldBe(FileMatcher.SearchAction.FailOnDriveEnumeratingWildcard);
+                    fileList.ShouldBeEmpty();
+                    excludeFileSpec.ShouldBe(string.Empty);
+
+                    (fileList, action, excludeFileSpec) = FileMatcher.Default.GetFiles(
+                        string.Empty,
+                        NativeMethodsShared.IsWindows ? @"\**" : "/**");
+
+                    action.ShouldBe(FileMatcher.SearchAction.FailOnDriveEnumeratingWildcard);
+                    fileList.ShouldBeEmpty();
+                    excludeFileSpec.ShouldBe(string.Empty);
+                }
+                finally
+                {
+                    ChangeWaves.ResetStateForTests();
+                }
+            }
+        }
+
+        [PlatformSpecific(TestPlatforms.Windows)]
+        [Theory]
+        [InlineData("c:\\", "**")]
+        [InlineData("c:\\\\", "**")]
+        [InlineData("c:\\\\\\\\", "**")]
+        [InlineData(@"c:\", @"**\*.cs")]
+        [InlineData(@"\", "**")]
+        [InlineData(@"c:////////", "**")]
+        public void TestDriveEnumeratingWildcardWindowsPattern(string directoryPart, string wildcardPart)
+        {
+            TestDriveEnumeratingWildcardPatternDetection(directoryPart, wildcardPart);
+        }
+
+        [PlatformSpecific(TestPlatforms.AnyUnix)]
+        [Theory]
+        [InlineData("/", "**/*.cs")]
+        [InlineData("/", "**")]
+        public void TestDriveEnumeratingWildcardUnixPattern(string directoryPart, string wildcardPart)
+        {
+            TestDriveEnumeratingWildcardPatternDetection(directoryPart, wildcardPart);
+        }
+
+        private void TestDriveEnumeratingWildcardPatternDetection(string directoryPart, string wildcardPart)
+        {
+            using (var env = TestEnvironment.Create())
+            {
+                ChangeWaves.ResetStateForTests();
+                env.SetEnvironmentVariable("MSBUILDFAILONDRIVEENUMERATINGWILDCARD", "1");
+                BuildEnvironmentHelper.ResetInstance_ForUnitTestsOnly();
+
+                try
+                {
+                    bool result = FileMatcher.IsDriveEnumeratingWildcardPattern(directoryPart, wildcardPart);
+                    result.ShouldBeTrue();
```

Comment:
Can you please also add some negative test cases where we expect `IsDriveEnumeratingWildcardPattern` to return false? Kind of related to my comment about detecting `Z:` on Windows-only because such patterns should return false on Unix.